ios/browser/api/bookmarks/exporter/bookmarks_encoder.cc

```diff
@@ -0,0 +1,191 @@
+  // NOTE: These functions take in individual properties of a bookmark node
+  // instead of taking in a BookmarkNode for efficiency so that we don't convert
+  // various data-types to UTF16 strings multiple times - once for serializing
+  // and once for computing the check-sum.
+  // The url parameter should be a valid UTF8 string.
+  void UpdateChecksumWithUrlNode(const std::string& id,
+                                 const base::string16& title,
+                                 const std::string& url);
+  void UpdateChecksumWithFolderNode(const std::string& id,
+                                    const base::string16& title);
+
+  // Initializes/Finalizes the checksum.
+  void InitializeChecksum();
+  void FinalizeChecksum();
+  
+  // MD5 context used to compute MD5 hash of all bookmark data.
+  base::MD5Context md5_context_;
+  
+  // Checksums.
+  std::string computed_checksum_;
+  std::string stored_checksum_;
+};
+
+BookmarkEncoder::BookmarkEncoder() = default;
+
+BookmarkEncoder::~BookmarkEncoder() = default;
+
+void BookmarkEncoder::UpdateChecksum(const std::string& str) {
+  base::MD5Update(&md5_context_, str);
+}
+
+void BookmarkEncoder::UpdateChecksum(const base::string16& str) {
+  base::MD5Update(&md5_context_,
+                  base::StringPiece(
+                      reinterpret_cast<const char*>(str.data()),
+                      str.length() * sizeof(str[0])));
+}
+
+void BookmarkEncoder::UpdateChecksumWithUrlNode(const std::string& id,
+                               const base::string16& title,
+                               const std::string& url) {
+  DCHECK(base::IsStringUTF8(url));
+  UpdateChecksum(id);
+  UpdateChecksum(title);
+  UpdateChecksum(bookmarks::BookmarkCodec::kTypeURL);
+  UpdateChecksum(url);
+}
+
+void BookmarkEncoder::UpdateChecksumWithFolderNode(const std::string& id,
+                                                 const base::string16& title) {
+  UpdateChecksum(id);
+  UpdateChecksum(title);
+  UpdateChecksum(bookmarks::BookmarkCodec::kTypeFolder);
+}
+
+void BookmarkEncoder::InitializeChecksum() {
+  base::MD5Init(&md5_context_);
+}
+
+void BookmarkEncoder::FinalizeChecksum() {
+  base::MD5Digest digest;
+  base::MD5Final(&digest, &md5_context_);
+  computed_checksum_ = base::MD5DigestToBase16(digest);
+}
+
+std::unique_ptr<base::Value> BookmarkEncoder::EncodeNode(const ExportedBookmarkEntry* node) {
+  std::unique_ptr<base::DictionaryValue> value(new base::DictionaryValue());
+  std::string id = base::NumberToString(node->id());
+  value->SetString(bookmarks::BookmarkCodec::kIdKey, id);
+  const base::string16& title = node->GetTitle();
+  value->SetString(bookmarks::BookmarkCodec::kNameKey, title);
+  const std::string& guid = node->guid();
+  value->SetString(bookmarks::BookmarkCodec::kGuidKey, guid);
+  value->SetString(bookmarks::BookmarkCodec::kDateAddedKey,
+                   base::NumberToString(node->date_added().ToInternalValue()));
```

Comment:
`ToInternalValue` is also not recommended: https://source.chromium.org/chromium/chromium/src/+/master:base/time/time.h;l=453?q=frominternalvalue&ss=chromium